python/lancedb/table.py

```diff
@@ -523,23 +524,109 @@ def update(
         raise NotImplementedError
 
 
+class LanceTableMode:
+    pass
+
+
+@dataclass
+class LanceStandardMode(LanceTableMode):
+    last_consistency_check: Optional[float] = None
+
+
+@dataclass
+class LanceTimeTravelMode(LanceTableMode):
+    version: int
+
+
 class LanceTable(Table):
     """
     A table in a LanceDB database.
+
+    This can be opened in two modes: standard and time-travel.
+
+    Standard mode is the default. In this mode, the table is mutable and tracks
+    the latest version of the table. The level of read consistency is controlled
+    by the `read_consistency_interval` parameter on the connection.
+
+    Time-travel mode is activated by specifying a version number. In this mode,
+    the table is immutable and fixed to a specific version. This is useful for
+    querying historical versions of the table.
     """
 
-    def __init__(self, connection: "LanceDBConnection", name: str, version: int = None):
+    def __init__(
+        self,
+        connection: "LanceDBConnection",
+        name: str,
+        version: Optional[int] = None,
+    ):
         self._conn = connection
         self.name = name
-        self._version = version
 
-    def _reset_dataset(self, version=None):
-        try:
-            if "_dataset" in self.__dict__:
-                del self.__dict__["_dataset"]
-            self._version = version
-        except AttributeError:
-            pass
+        if version is not None:
+            self._mode = LanceTimeTravelMode(version)
+        else:
+            self._mode = LanceStandardMode()
+
+        # Dataset is lazily loaded
+        self._ds: Optional[LanceDataset] = None
+
+    @classmethod
+    def open(cls, db, name, **kwargs):
+        tbl = cls(db, name, **kwargs)
+        fs, path = fs_from_uri(tbl._dataset_uri)
+        file_info = fs.get_file_info(path)
+        if file_info.type != pa.fs.FileType.Directory:
+            raise FileNotFoundError(
+                f"Table {name} does not exist."
+                f"Please first call db.create_table({name}, data)"
+            )
+        register_event("open_table")
+
+        return tbl
+
+    @property
+    def _dataset_uri(self) -> str:
+        return join_uri(self._conn.uri, f"{self.name}.lance")
+
+    @property
+    def _dataset(self) -> LanceDataset:
+        # Returns the LanceDataset this wraps. This handles lazy loading (if
+        # self._ds is None) and checking for new versions (if user has specified
+        # a read_consistency_interval).
+        if not self._ds and isinstance(self._mode, LanceTimeTravelMode):
```

Comment:
Instead of calling it a "mode", what if we called this `LanceDatasetRef` and you have immutable ("time travel mode") and mutable ("standard mode") variants. You construct a reference by passing in the dataset uri, the version (if specified), and the read_consistency_interval. The Ref holds on to the `._ds` dataset.

And the Ref class can have `get_dataset()` and `get_dataset_mut()` methods, and for the immutable one it just raises a TypeError.

This basically allows us to avoid all the different isinstance checks and don't need to keep all branches of the cascading if straight in our heads.

wdyt?

Reply:
I like that idea. I can change to match that.